src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,331 @@
+	eg.Go(func() error { return ring.watch(ctx, cancel, etcdCol) })
+	eg.Go(func() error { return ring.createNode(ctx, cancel, etcdCol, ring.node.Id) })
+	eg.Go(func() error {
+		if err := cb(ctx, ring); err != nil {
+			return err
+		}
+		cancel()
+		return nil
+	})
+	err := eg.Wait()
+	if errors.Is(cancelCtx.Err(), context.Canceled) {
+		err = nil
+	}
+	return errors.EnsureStack(err)
+}
+
+// ring is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func ring(client *etcd.Client, logger *logrus.Logger, prefix string, id string) *Ring {
+	localMember := member{
+		Id:   id,
+		hash: hashFn([]byte(id)),
+	}
+	return &Ring{
+		client:  client,
+		hashFn:  hashFn,
+		logger:  logger,
+		members: []member{localMember},
+		node: node{
+			member: localMember,
+			locks:  map[string]lockInfo{},
+		},
+		prefix: prefix,
+	}
+}
+
+// createNode creates a lease, inserts itself as a key to etcd, keeps the lease alive in the background.
+func (ring *Ring) createNode(ctx context.Context, cancel context.CancelFunc, col collection.EtcdCollection, id string) error {
+	if err := col.Claim(ctx, id, &types.BoolValue{Value: true},
+		func(ctx context.Context) error {
+			// keep the lease alive until the context is canceled.
+			<-ctx.Done()
+			return errors.EnsureStack(ctx.Err())
+		}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			return nil
+		}
+		cancel()
+		ring.logger.WithFields(ring.memberLogFields()).Errorf("failed to renew etcd lease: %v", err)
+		return errors.EnsureStack(err)
+	}
+	return nil
+}
+
+// watch watches etcd keys with prefix Ring.prefix to determine if a node has been added or removed.
+// When a node is added, the ring calls rebalance to release any locks that no longer associate to its node. Rebalance
+// on delete happens by default since each member attempts to retrieve all locks. When a member is deleted,
+// the pending calls to Lock by each member will go through on the nodes that associates to given lock.
+func (ring *Ring) watch(ctx context.Context, cancel context.CancelFunc, col collection.EtcdCollection) error {
+	if err := col.ReadOnly(ctx).WatchF(func(event *watch.Event) error {
+		ring.stateLock.Lock()
+		defer ring.stateLock.Unlock()
+		nodeKey := string(event.Key)
+		switch event.Type {
+		case watch.EventDelete:
+			ring.removeById(ring.etcdToMember(nodeKey))
+		case watch.EventPut:
+			ring.insertById(ring.etcdToMember(nodeKey))
+			// Need to release locks that no longer associate to the ring's node.
+			if err := ring.rebalance(); err != nil {
+				ring.logger.WithField("ring", ring.prefix).Errorf("failed rebalancing: %v", err)
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			ring.logger.WithFields(ring.memberLogFields()).Infof("shutting down ring")
```

Comment:
This should probably just be in a defer in the with ring function. Also, a setup log may be nice as well.